### DISCUSSION TITLE
Chained events in a greeting

### DISCUSSION THREAD

**Discussion by jamalainm** [evennia/evennia#2469] [Evennia Questions] [answered]
Hi all,

I'd like to have an NPC greet a newcomer and after a few seconds say something else to the newcomer and then send a message to the newcomer with some helpful notes for people new to the world of MUDs.

I started by adding a callback to the NPC:

```
character.execute_cmd(f"say Hi, {newcomer}!")
call_event(character, "chain_1", 1)
```

And then adding chain_1 to the NPC:

`character.execute_cmd("say How can I help you?")`

However, chain_1 ends up making the newcomer do the speaking, not the NPC, who I would expect to be designated by the `character` variable.

Additionally, I wanted to add some instructional material to the newcomer, so I added to the greet callback:

`call_event(newcomer, "chain_2", 2)`

And I added chain_2 to the NPC:

`newcomer.msg("NOTE: to talk, type 'say'")`

This message is not delivered to the newcomer and does not issue any error message to the admin.

Apologies if I'm missing something basic or if there's a better way to have delayed actions or behaviors.

Thanks for your time! Take care!

**Comment by kovitikus** [comment, marked as answer]
Hello,

You should use [`evennia.utils.utils.delay()`](https://github.com/evennia/evennia/blob/master/evennia/utils/utils.py#L1023) for code that you want to execute in X amount of seconds. It's a pretty straight-foward method.

As far as what object is stored in character, it depends on what your code looks like where you set this. It also depends on where you are writing this code from (what class is the self?). Keep in mind that when you are coding, you are generally doing so from a perspective and it's best to track needed objects through a variable, which you can then reference.

**Reply by jamalainm** [reply]
Ah! I had been trying to get this going with the in-game python, where the greet event itself designates the identity of `character` and `newcomer`.

Are you thinking it would be easier / more straight forward to write up the NPC as its own typeclass? If I were to do that, do you have thoughts about how to make an NPC listen for a character's arrival? I'm looking at the [tutorial for NPC's listening](https://evennia.readthedocs.io/en/latest/Tutorial:--NPC's-listening.html?highlight=NPC#tutorial-npcs-listening), but I'm not sure whether/how I could modify it for my purposes.

Thanks for the advice! Take care!

**Reply by davewiththenicehat** [reply]
https://www.evennia.com/docs/latest/Tutorial-Aggressive-NPCs.html
Covers `Character` method `at_char_entered`
I think that is what you are looking for.

**Reply by jamalainm** [reply]
Perfect! Thanks all!
